### PR TITLE
Crusher Rework

### PR DESCRIPTION
**Pull request by Surrealaser**
-Crusher gains a new ability: Crest Fling. This will fling a mob, (including friendly smaller Xenos) 3-5 tiles behind or in front of the Crusher, scattering up to 1 tile away.
-While on Harm intent, Crest Fling will throw the target in front of the Crusher, otherwise, it will throw the target behind the Crusher.
-Deals minor Brute and moderate Halloss damage to non-Xenos that scales with the distance the target was flung. Halloss damage ignores armour.
-This ability has a cooldown of 6 seconds and costs 40 plasma.

-Stomp reworked. It now affects mobs within 2 tiles of the Crusher, dealing base damage equal to 1.5x the Crusher's base melee damage in Halloss and Brute.
-All damage dealt by Stomp is divided by the distance the target is from the Crusher. 
-Stomp Brute damage now subject to armor. All Halloss damage inflicted in this way ignores armor.
-Knocks targets within Range 1 prone for 1 tick; stuns all other targets for 1 tick. 
-Targets at Range 1 and 2 knocked 1 tile away.
-Now costs 80 plasma up from 50.

-Crusher rear/side armour reduced by 75% of its base armour, up from 50% of its base armour.

-Residual damage application functions changed to utilize new fixed % reduction.

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 1]
```diff
@@ -1468,3 +1468,96 @@
...
+		if(new_target)
+			T = new_target//Looks like we found a turf.
+
+		M.throw_at(T, toss_distance, 1, src)
+		M.KnockDown(1, 1)
+
+		//Handle the damage
+		if(!isXeno(M)) //Friendly xenos don't take damage.
+			var/damage = toss_distance * 5
+			if(frenzy_aura > 0)
```
Comment (Qustinnus) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: can't you just do if(frenzy_aura)

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 2]
```diff
@@ -1468,3 +1468,96 @@
...
+				damage *= (1 + round(frenzy_aura * 0.1,0.01))
+			var/armor_block = M.run_armor_check("chest", "melee")
+			if(ishuman(M))
+				var/mob/living/carbon/human/H = M
+				H.take_overall_damage(rand(damage * 0.75,damage * 1.25) * 0.5, armor_block) //Armour functions against this.
```
Comment (Qustinnus) [inline code/modules/mob/living/carbon/xenomorph/Powers.dm]: Redundant typecheck

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 3]
```diff
@@ -79,7 +79,14 @@
 
 
 /mob/living/carbon/Xenomorph/apply_damage(damage = 0, damagetype = BRUTE, def_zone = null, blocked = 0, used_weapon = null, sharp = 0, edge = 0)
-	if(!damage) return
+	if(blocked >= 1) //total negation
+		return 0
```
Comment (Qustinnus) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm]: Defineees

File: code/modules/mob/living/carbon/xenomorph/Castes/Crusher.dm [hunk 4]
```diff
@@ -63,21 +68,33 @@
...
+		var/distance = get_dist(M, loc)
+		var/damage = (rand(melee_damage_lower, melee_damage_upper) * 1.5) / max(1,distance + 1)
+		if(frenzy_aura > 0)
+			damage *= (1 + round(frenzy_aura * 0.1,0.01)) //+10% per level of Frenzy
+		if(distance == 0) //If we're on top of our victim, give him the full impact
+			round_statistics.crusher_stomp_victims++
+			var/armor_block = M.run_armor_check("chest", "melee")
+			if(ishuman(M))
+				var/mob/living/carbon/human/H = M
+				H.take_overall_damage(rand(damage * 0.75,damage * 1.25), armor_block) //Armour functions against this.
```
[the remark this answers is not in the document]
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Castes/Crusher.dm]: The human version of the function is different and has more advanced limb handling.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm [hunk 5]
```diff
@@ -1468,3 +1468,96 @@
...
+
+	if(stagger)
+		to_chat(src, "<span class='xenowarning'>You try to fling away [M] but are unable as you fail to shake off the shock!</span>")
+		return
+
+	if (!Adjacent(M)) //Sanity check
+		return
+
+	if(!istype(M, /mob/living)) //Sanity check
+		return
```
[the remark this answers is not in the document]
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: Only the ones that don't have their own special notifications; I'll consolidate as much of them as I can.

File: code/modules/mob/living/carbon/xenomorph/Powers.dm (hunk 2 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/Powers.dm]: As above, these procs work differently for non-human mobs vs humans.

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm [hunk 6]
```diff
@@ -1198,6 +1198,19 @@
 		button.overlays.Cut()
 		button.overlays += image('icons/mob/actions.dmi', button, "savage_on")
 
+// Defender Headbutt
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/Abilities.dm]: Headbutt?

**Comment by Rohesie** [comment]
Stomp dealing area halloss might be a bit bad, I'll need more opinions on this.

**Pull request merged**